lib/uniform/blueprint.ex

```diff
@@ -7,25 +7,24 @@ defmodule Uniform.Blueprint.BeforeCompile do
       if templates_dir do
         templates_dir
         |> Path.join("**/*.eex")
-        |> Path.wildcard()
+        |> Path.wildcard(match_dot: true)
       else
         []
       end
 
     template_functions =
       for path <- templates do
+        quoted = EEx.compile_file(path)
         path = Path.relative_to(path, templates_dir)
 
         quote do
           @file unquote(path)
           @external_resource unquote(path)
 
-          EEx.function_from_file(
-            :def,
-            unquote(String.to_atom(path)),
-            unquote(Path.join(templates_dir, path)),
-            [:app]
-          )
+          def unquote(String.to_atom(path))(var!(app)) do
+            _ = var!(app)
+            unquote(quoted)
+          end
```

Comment:
These four lines mimic [Phoenix.View](https://github.com/phoenixframework/phoenix_view/blob/79ed7e428270f7b150c71990d92302db34f2672f/lib/phoenix/template.ex#L426-L429).

---

`def unquote(String.to_atom(path))` starts creating a function. If `path` were `"foo"` you'd end up with `def foo`. (Function names are atoms.)

However, it actually ends up creating functions where the name of the function is a path to a template. For example `def :"config/runtime.exs"`. (That isn't valid Elixir code; it won't compile. But it's the name you end up with.)

---

`(var!(app))` turns into `(app)`. As in `def foo(app)`. `var!` is about telling the compile a variable in a quoted expression should not be hygienized. For some reason related to how we're building the function, this is required. (Things break if you remove it or replace with `Macro.var`.) I'm doing a bit of conscious handwaving here because I don't completely understand all of the concepts involved.

---

`_ = var!(app)` is the bit that removes the warning about the unused variable.

Compare with this:

```elixir
defmodule A do
  def b(c) do
    :ok
  end
end

warning: variable "c" is unused (if the variable is not meant to be used, prefix it with an underscore)
  iex:2: A.b/1
```

```elixir
defmodule Z do
  def y(x) do
    _ = x
    :ok
  end
end

[no warning]
```

By "using" the variable – even if you set it to another throwaway variable (`_`) – the compiler considers it used and stops complaining.

---

`unquote(quoted)` (paired with `quoted = EEx.compile_file(path)` above) injects the AST of some code that expects the `app` variable to exist and evaluates to the rendered template. Essentially, the body of a function to render the template.